### PR TITLE
[codex] TASK-050 high-priority security remediation

### PR DESCRIPTION
**Pull request by dorianagaesse**
## What changed
Implemented the ranked remediation slice from TASK-049 across the public auth and agent-access perimeter.

This PR:
- adds DB-backed abuse controls for sign-in, sign-up, forgot-password, verification resend, and `POST /api/auth/agent/token`
- records failed sign-in and failed agent token exchange activity without storing raw secrets
- hashes human session tokens at rest and switches session lookup/revocation to hashed-token semantics
- invalidates legacy plaintext-backed human sessions during migration instead of keeping indefinite dual-format compatibility
- makes already-issued agent bearer tokens honor current credential liveness during request-time usage checks
- records the architecture choices in `adr/decisions.md`, a dedicated TASK-050 ADR, `journal.md`, and the current task brief

## Why
TASK-049 identified three top remaining gaps:
1. no real public auth/token abuse-control baseline
2. plaintext human sessions at rest
3. agent bearer tokens staying valid until TTL even after rotate/revoke

This PR closes those three findings without expanding into unrelated hardening scope.

## Impact
- public auth and token exchange now fail closed on obvious overuse with bounded metadata and stable user-facing responses
- database reads alone no longer expose active human session tokens
- rotating or revoking an agent credential now cuts off already-issued bearer-token use immediately at request time
- one-time rollout effect: existing human sessions are invalidated by the session-token migration

## Validation
- `npm run lint`
- `npm test`
- `npm run test:coverage`
- `npx prisma generate`
- `npm run build` with safe preview-style env overrides

## Known local blocker
- `npm run test:e2e` remains blocked in this workstation session because Prisma cannot reach the expected PostgreSQL fixture service at `127.0.0.1:5432`
- CI/preview should provide the authoritative e2e signal for this branch

### CODE REVIEW

File: prisma/migrations/20260410110000_task050_security_remediation/migration.sql [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+);
+
+CREATE TABLE "AuthRateLimitBucket" (
+  "scope" "AuthRateLimitScope" NOT NULL,
+  "key" VARCHAR(191) NOT NULL,
+  "windowStart" TIMESTAMPTZ(6) NOT NULL,
+  "attemptCount" INTEGER NOT NULL DEFAULT 0,
+  "blockedUntil" TIMESTAMPTZ(6),
+  "createdAt" TIMESTAMPTZ(6) NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  "updatedAt" TIMESTAMPTZ(6) NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
[the remark this answers is not in the document]
Comment (dorianagaesse) [reply prisma/migrations/20260410110000_task050_security_remediation/migration.sql]: Aligned the migration with the repo’s existing Prisma pattern in commit `c4d1dea`: `AuthRateLimitBucket.updatedAt` is now `NOT NULL` without a DB default, and Prisma remains the source of truth for updates.

File: lib/services/auth-abuse-control-service.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            THEN ${blockedUntil}
+          ELSE "AuthRateLimitBucket"."blockedUntil"
+        END,
+        "updatedAt" = CURRENT_TIMESTAMP
+      RETURNING "blockedUntil"
+    `
+  );
+
+  return rows[0] ?? { blockedUntil: null };
+}
```
[the remark this answers is not in the document]
Comment (dorianagaesse) [reply lib/services/auth-abuse-control-service.ts:186]: Added a bounded retention sweep in commit `c4d1dea`. `consumeAuthAbuseQuota` now prunes expired buckets for the participating scope/key set, keeping a small retention window while preserving active blocks.

File: lib/services/credential-auth-service.ts [hunk 3]
```diff
@@ -210,14 +397,31 @@ export async function signInWithEmailPassword(input: {
...
+      requestId: input.requestId,
+      userAgent: input.userAgent,
+      reason: "password-mismatch",
+    });
   }
 
+  await clearAuthAbuseControls({
+    scope: AuthRateLimitScope.sign_in,
+    keys: abuseSignals.map((signal) => signal.key),
+  });
```
[the remark this answers is not in the document]
Comment (dorianagaesse) [reply lib/services/credential-auth-service.ts]: Adjusted in commit `c4d1dea`. Successful sign-in now treats limiter cleanup as best-effort: cleanup failures log a warning but no longer fail an otherwise valid authentication.

File: lib/services/project-agent-access-service.ts [hunk 4]
```diff
@@ -854,21 +1029,59 @@ export async function exchangeAgentApiKeyForAccessToken(input: {
...
+      userAgent: input.userAgent,
+      reason: credential.revokedAt ? "credential-revoked" : "credential-expired",
+    });
+    return createError(failureResult.ok ? 401 : 429, failureResult.ok ? "invalid-api-key" : "too-many-attempts");
   }
 
+  await clearAuthAbuseControls({
+    scope: AuthRateLimitScope.agent_token_exchange,
+    keys: abuseSignals.map((signal) => signal.key),
+  });
```
[the remark this answers is not in the document]
Comment (dorianagaesse) [reply lib/services/project-agent-access-service.ts]: Adjusted in commit `c4d1dea`. Successful agent token exchange now treats limiter cleanup as best-effort and logs a bounded warning instead of failing the token issue path.

File: lib/services/project-agent-access-service.ts [hunk 5]
```diff
@@ -306,9 +319,120 @@ function buildAuditMetadata(input: {
...
+  userAgent?: string | null;
+  reason: string;
+}): Promise<void> {
+  logServerWarning(
+    "projectAgentAccess.tokenExchangeFailed",
+    "Agent token exchange failed.",
+    {
+      requestId: input.requestId ?? null,
+      ipAddress: input.ipAddress ?? null,
+      userAgent: input.userAgent ?? null,
```
[the remark this answers is not in the document]
Comment (dorianagaesse) [reply lib/services/project-agent-access-service.ts]: Addressed in commit `c4d1dea`. Failed token-exchange warning logs now normalize and bound `ipAddress` / `userAgent` before emitting metadata, matching the audit-path safeguards.

File: lib/services/credential-auth-service.ts [hunk 6]
```diff
@@ -114,13 +127,153 @@ function isUsernameDiscriminatorConstraint(error: unknown): boolean {
...
+
+  logServerWarning("credentialAuth.signInFailed", "Email/password sign-in failed.", {
+    requestId: input.requestId ?? null,
+    ipAddress: input.ipAddress ?? null,
+    userAgent: input.userAgent ?? null,
+    reason: input.reason,
+    accountKey: buildAuthRateLimitKey("sign-in:email", input.email),
+    throttled: !abuseResult.ok,
+    retryAfterSeconds: abuseResult.ok ? null : abuseResult.retryAfterSeconds,
+  });
```
[the remark this answers is not in the document]
Comment (dorianagaesse) [reply lib/services/credential-auth-service.ts:264]: Addressed in commit `c4d1dea`. Credential-auth warning logs now bound attacker-controlled `ipAddress` / `userAgent` metadata before logging, and the success-path cleanup warning follows the same rule.

File: lib/http/request-metadata.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    const firstAddress = forwardedFor
+      .split(",")
+      .map((entry) => entry.trim())
+      .find((entry) => entry.length > 0);
+    if (firstAddress) {
+      return firstAddress;
+    }
+  }
+
+  return headers.get("x-real-ip")?.trim() ?? null;
```
[the remark this answers is not in the document]
Comment (dorianagaesse) [reply lib/http/request-metadata.ts]: Fixed in commit `c4d1dea`. `readClientIpAddressFromHeaders` now normalizes blank `x-real-ip` values to `null`, and I added a focused regression test for that edge case.
